### ISSUE TITLE
Document/automate how to migrate from a Wrangler 1 `webpack_config` / webpack integration to Wrangler 2

### ISSUE THREAD

**Issue by threepointone** [cloudflare/wrangler2#8]
wrangler v2 deprecates the `type` and `webpack_config` configurations, and it's overall webpack integration story. 

- This has historically been a pain point both internally and externally. We haven't been able to update the webpack version (v4) because of breaking changes, and it scares away people from using wrangler in the first place. 
- The other two values for `type` - `javascript` and `rust` don't make much sense anymore since it's really the same thing backing it. And we're going to encourage development in multiple languages (Go, python, what have you) since we want to have the best `wasm` support ever. 
- Because we now ship esbuild, we can automatically provide support for a number of features  that bundlers are typically used for: modules, jsx, typescript, etc. 

TODO: as part of the effort to be backward compatible, we should have a super low-effort transition guide to moving existing webpack projects to the new wrangler (ideally, we should be able to automate it with a single command/codemod).

**Comment by touhidurrr** [comment]
So, since webpack_config is deprecated, what do I use?

**Comment by touhidurrr** [comment]
I want wrangler use webpack on production mode

**Comment by petebacondarwin** [comment]
In Wrangler 2, we use esbuild to bundle the code into a single file (more or less) for uploading to Cloudflare.
So for the majority of cases this completely replaces the webpack bundling in Wrangler 1.
But if you really need to use webpack then our recommended approach is to add it as a prepublish step. There a many ways to implement that but here is one...

In `package.json` use:

```
"scripts" {
  "publish": "webpack && wrangler publish",
  "dev": "concurrently -p --kill-others \"webpack --watch\" \"wrangler dev\""
}
```

**Comment by touhidurrr** [comment]
> In Wrangler 2, we use esbuild to bundle the code into a single file (more or less) for uploading to Cloudflare. So for the majority of cases this completely replaces the webpack bundling in Wrangler 1. But if you really need to use webpack then our recommended approach is to add it as a prepublish step. There a many ways to implement that but here is one...
> 
> In `package.json` use:
> 
> ```
> "scripts" {
>   "publish": "webpack && wrangler publish",
>   "dev": "concurrently -p --kill-others \"webpack --watch\" \"wrangler dev\""
> }
> ```

Never heard about `esbuild` before. I have some questions.
About `esbuild`....
1. Does it have Some kind of command that gives the same functionality as `webpack`'s production mode?
2.  Do you guys allow config customization for `esbuild`?
About the default behaviour of `wrangler2`...
1.  What is the default index file  for `wrangler2`? `wrangler2 publish` throws errors.
2. How to set default `wrangler2` index file. Previously custom webpack config file was used I think.
3. Can I you guys make `wrangler2` save is output file? Like a customizable output path for testing it’s behaviour.
About development Pipeline...
1. I want to use Github Workflows for publishing my project. What Authentication method do you suggest? Surely `wrangler2 login` isn't an option. Or do you suggest Cloudflare API?

**Comment by petebacondarwin** [comment]
> Does it have Some kind of command that gives the same functionality as webpack's production mode?

The output of esbuild is a bundled file where imports have been inlined (mostly) similar to webpack. Currently we are not minifying the output. What production mode features of webpack are you looking for here?

> Do you guys allow config customization for esbuild?

For now we are treating the bundler as an internal implementation aspect of Wrangler 2. So at the moment we are not allowing the bundling to be extended. There is an issue to track this: https://github.com/cloudflare/wrangler2/issues/234.

> What is the default index file for wrangler2?

There is no default index file for Wrangler 2. It must be configured...

> How to set default wrangler2 index file.

You can specify the entry-point index file for a Worker in a variety of ways:

- In `wrangler publish` and `wrangler dev` the first positional command line argument is the path to the index file.
- If not provided at the command line, you can specify it in the wrangler.toml via the `main` property. This is tracked here: https://github.com/cloudflare/wrangler2/issues/575.

> Can I you guys make wrangler2 save is output file? Like a customizable output path for testing it’s behaviour.

Currently we do not have a facility to save the output from the bundling to disk.
We would like to improve the testing and debugging story for Workers. There is this issue about improving debugging: https://github.com/cloudflare/wrangler2/issues/573

I want to use Github Workflows for publishing my project. What Authentication method do you suggest? Surely wrangler2 login isn't an option. Or do you suggest Cloudflare API?

If you are publishing from a CI where you cannot have an interactive login, you can currently specify an API token via the `CLOUDFLARE_API_TOKEN` environment variable.

**Comment by touhidurrr** [comment]
Ok also I read `esbuild` docs. If you are using `esbuild` treating as an internal package then I guess it doesn’t honor `esbuild` features like `browser` parameter in `package.json`. Or, does it do since it’s being executed from the same path as the `package.json` file?

**Comment by petebacondarwin** [comment]
We pass all the options to esbuild programmatically, so it should not be reading anything from package.json. (And if it is then that is a bug!)

**Comment by caass** [comment]
@touhidurrr this should be fixed via #759, can you let us know if the webpack plugin works as expected?

**Comment by touhidurrr** [comment]
> @touhidurrr this should be fixed via #759, can you let us know if the webpack plugin works as expected?

Dunno actually. Already moved to `esbuild` cause the build time is much faster than webpack. 😅

**Comment by threepointone** [comment]
Closing this anyway. Thank you @caass!
